audformat/core/utils.py

```diff
@@ -353,6 +353,36 @@ def hash(
     return str(pd.util.hash_pandas_object(obj).sum())
 
 
+def index_has_overlap(
+        obj: typing.Union[pd.Index, pd.DataFrame, pd.Series],
+) -> bool:
+    r"""Check if one or more segments in the index overlap.
+
+    Args:
+        obj: object conform to
+            :ref:`table specifications <data-tables:Tables>`
+
+    Returns:
+        ``True`` if overlap is detected, otherwise ``False``
+
+    """
+
+    index = obj if isinstance(obj, pd.Index) else obj.index
+
+    if index_type(index) == define.IndexType.FILEWISE:
+        return False
+
+    index = to_segmented_index(index, allow_nat=False)
```

Comment:
Mhh, I'm not sure if we should force `allow_nat=False` here as it can be costly. Instead we should probably handle `NaT`. E.g.:

```
audformat.segmented_index(
    ['f1'] * 2,
    [0, 2],
    [1, NaT],
)
```

-> False

```
audformat.segmented_index(
    ['f1'] * 2,
    [0, 2],
    [NaT, 3],
)
```

-> True

```
audformat.segmented_index(
    ['f1'] * 2,
    [0, 2],
    [NaT, NaT],
)
```

-> True

Reply:
Hm why and how costly it is ?

Reply:
It will replace `NaT` with the duration of the file, so it has to get the duration for every file that is affected.

Reply:
In fact, it's not necessary to replace `NaT` with an exact value anyway. The result will be the same as long as we replace it with a large enough value. So I recommend to remove the line and instead insert:

```
ends = ends.fillna(pd.to_timedelta(sys.maxsize))
```

before the `if` condition within the loop.

Reply:
Yes I agree this is smarter way to do it. I applied the suggestions and added the adequate tests.

Reply:
By the way from where can I get the knowledge of what's faster and what's not and these kind of optimizations when handling data structures ? I mean I find things on internet from time to time but maybe you have a better source

Reply:
Any operation that needs to access information from disk becomes costly once you scale it up. Otherwise it's not always obvious what will be costly and what not. So usually the only way to find a bottleneck is by running your function on a large number of files / segments. With `audb` we also used a profiler at some point to nail down the expensive operations.

Reply:
Hm okay, a profiler is a tool that helps you simulate bottlenecks and gives you statistics and stuff right ? And what did you once you figured what are the expensive operations ? Some optimizations or just noting down that they should be avoided on large structures ?

Reply:
> Hm okay, a profiler is a tool that helps you simulate bottlenecks and gives you statistics and stuff right ?

Yes, a profiler shows you how long each function call takes and helps you to identify the critical calls

> And what did you once you figured what are the expensive operations?

Try to optimize you code by avoiding expensive operations :) E.g. here we now replace `NaT` with a large value which is much faster than accessing the actual file duration. But the result will be the same. These kind of things.

Reply:
Okay perfect, thanks for explaining. Feel free to resolve the thread.